dom.bs

```diff
@@ -2652,6 +2652,50 @@ of a <var>node</var> into a <var>parent</var> before a <var>child</var>, run the
  <!-- Technically this is post-insert. -->
 </ol>
 
+<p>To <dfn export for=Node id=concept-node-ensure-pre-move-validity>ensure pre-move validity</dfn>
+of a <var>node</var> into a <var>parent</var> before a <var>child</var>, run these steps:
+
+<ol>
+ <li><p>If either <var>parent</var> or <var>node</var> are not <a>connected</a>, then
+ <a>throw</a> a "{{HierarchyRequestError!!exception}}" {{DOMException}}.</p></li>
+
+ <li><p>If <var>parent</var>'s <a for=/>shadow-including root</a> is not the same as
+ <var>node</var>'s <a for=/>shadow-including root</a>, then <a>throw</a> a
+ "{{HierarchyRequestError!!exception}}" {{DOMException}}.</p></li>
+
+ <li><p>If <var>parent</var> is not an {{Element}} <a for=/>node</a>, then <a>throw</a> a
+ "{{HierarchyRequestError!!exception}}" {{DOMException}}.
```

Comment:
Why moves in DocumentFragments aren't possible? Especially, what if parent is a ShadowRoot?

Reply:
DocumentFragments are always disconnected, right? We require that the moved node, and its new parent, are both connected to the DOM.

Reply:
Shadowroot is a DocumentFragment. I would expect moves under ShadowRoot being possible.
(I wasn't thinking the connectedness case here, only that one should be able reorder the children of a ShadowRoot)

Reply:
I didn't realize `ShadowRoot` was a `DocumentFragment`! Thanks for pointing this out. OK I've changed the spec to enable this (we can discuss connected-ness separately), and I've added a test for this in https://crrev.com/c/6012161. I'll resolve this now.